lib/api/objectDelete.js

```diff
@@ -24,48 +28,124 @@ export default function objectDelete(authInfo, request, log, cb) {
+        },
+        callback => services.preprocessingVersioningDelete(bucketName,
+            bucketMD, objectKey, objectMD, reqVersionId, log, callback),
+        (options, callback) => {
+            if (options && options.deleteData) {
+                // delete object
+                deleteOptions = options;
+                return callback(null, options);
             }
-            if (objMD['content-length']) {
-                log.end().addDefaultFields({
-                    contentLength: objMD['content-length'],
-                });
+            // putting a new delete marker
+            if (bucketMD.hasDeletedFlag() &&
+                    canonicalID !== bucketMD.getOwner()) {
+                log.trace('deleted flag on bucket and request ' +
+                        'from non-owner account');
+                return callback(errors.NoSuchBucket);
             }
-            return services.deleteObject(bucketName, objMD, objectKey, log,
-                err => {
-                    if (err) {
-                        return cb(err, corsHeaders);
-                    }
-                    pushMetric('deleteObject', log, {
-                        authInfo,
-                        bucket: bucketName,
-                        byteLength: objMD['content-length'],
-                        numberOfObjects: 1,
-                    });
-                    return cb(null, corsHeaders);
-                });
-        });
+            if (bucketMD.hasTransientFlag() || bucketMD.hasDeletedFlag()) {
+                return cleanUpBucket(bucketMD, canonicalID,
+                        log, err => callback(err, null));
+            }
+            return callback(null, null);
+        },
+        (options, callback) => {
+            if (options && options.deleteData) {
+                if (objectMD.isDeleteMarker) {
+                    removeDeleteMarker = true;
+                }
+                return services.deleteObject(bucketName, objectMD, objectKey,
+                        options, log, callback);
+            }
+            request.isDeleteMarker = true; // eslint-disable-line
+            return createAndStoreObject(bucketName, bucketMD,
+                objectKey, objectMD, authInfo, canonicalID, null, request,
+                null, log, callback);
+        },
+    ], (err, res) => {
+        if (err) {
+            log.debug('error processing request', { error: err,
+                method: 'metadataValidateAuthorization' });
+        } else if (deleteOptions === undefined) {
+            // TODO metric for delete marker
+            if (res.versionId) {
+                corsHeaders['x-amz-delete-marker'] = true;
+                corsHeaders['x-amz-version-id'] = res.versionId === 'null' ?
+                    res.versionId : VID.encrypt(res.versionId);
+            }
+            pushMetric('putObject', log, { authInfo, bucket: bucketName,
+                newByteLength: 0, oldByteLength: 0 });
```

Comment:
can you check with @rahulreddy which behavior we should have here, putting the old length ?